rs/rosetta-api/icrc1/ledger/tests/tests.rs

```diff
@@ -1543,3 +1553,154 @@ mod verify_written_blocks {
+            accounts_overflow_trim_quantity: None,
+        }))
+        .unwrap()
+    }
+
+    #[test]
+    fn should_successfully_upgrade_ledger_from_u64_to_u256_to_u64_wasm() {
+        let env = StateMachine::new();
+        let ledger_id = env
+            .install_canister(ledger_mainnet_u64_wasm(), default_init_args(), None)
+            .unwrap();
+        // Create a large balance
+        transfer(&env, ledger_id, MINTER, account(1), u64::MAX);
+        let mut balance = balance_of(&env, ledger_id, account(1));
+        // Create a large allowance
+        let approval_result = send_approval(
+            &env,
+            ledger_id,
+            account(1).owner,
+            &ApproveArgs {
+                from_subaccount: None,
+                spender: account(2),
+                amount: Nat::from(u64::MAX),
+                expected_allowance: None,
+                expires_at: None,
+                fee: None,
+                memo: None,
+                created_at_time: None,
+            },
+        );
+        assert_eq!(approval_result, Ok(BlockIndex::from(1u64)));
+        balance -= FEE;
+
+        // Try to upgrade the ledger from using a u64 wasm to a u256 wasm
+        let upgrade_args = Encode!(&LedgerArgument::Upgrade(None)).unwrap();
+        env.upgrade_canister(ledger_id, ledger_mainnet_u256_wasm(), upgrade_args)
+            .expect("Unable to upgrade the ledger canister");
+
+        // The balance and allowance should not change
+        assert_eq!(balance, balance_of(&env, ledger_id, account(1)));
+        let allowance = get_allowance(&env, ledger_id, account(1), account(2));
+        assert_eq!(allowance.allowance, Nat::from(u64::MAX));
+
+        // Try to upgrade the ledger back to a u64 wasm
+        let upgrade_args = Encode!(&LedgerArgument::Upgrade(None)).unwrap();
+        env.upgrade_canister(ledger_id, ledger_mainnet_u64_wasm(), upgrade_args)
+            .expect("Unable to upgrade the ledger canister");
+
+        // The balance and allowance should not change
+        assert_eq!(balance, balance_of(&env, ledger_id, account(1)));
+        let allowance = get_allowance(&env, ledger_id, account(1), account(2));
+        assert_eq!(allowance.allowance, Nat::from(u64::MAX));
+    }
+
+    #[test]
+    fn should_fail_to_upgrade_to_u64_ledger_wasm_if_total_supply_too_large() {
+        let env = StateMachine::new();
+        let ledger_id = env
+            .install_canister(ledger_mainnet_u256_wasm(), default_init_args(), None)
+            .unwrap();
+        transfer(&env, ledger_id, MINTER, account(1), u64::MAX - 100);
+        transfer(&env, ledger_id, MINTER, account(2), u64::MAX - 100);
+        let balance_1 = balance_of(&env, ledger_id, account(1));
+        let balance_2 = balance_of(&env, ledger_id, account(2));
+        assert_eq!(balance_1, balance_of(&env, ledger_id, account(1)));
+        assert_eq!(balance_2, balance_of(&env, ledger_id, account(2)));
```

Comment:
Are these asserts needed?

Reply:
Hah, yes, good catch! I did want to include them to make sure that total supply is larger than `u64::MAX`. I changed them to what you suggested.